--- original/pptp.c	2008-05-13 23:33:55.000000000 -0700
+++ ./pptp.c	2012-01-25 10:25:15.141070369 -0800
@@ -59,7 +59,7 @@
 #include "pptp_callmgr.h"
 #include "pptp_gre.h"
 #include "version.h"
-#if defined(__linux__)
+#if PR_SET_NAME
 #include <sys/prctl.h>
 #else
 #include "inststr.h"
--- pptp_callmgr.original	2012-02-01 17:03:14.568770519 -0800
+++ pptp_callmgr.c	2012-02-01 17:03:19.376660239 -0800
@@ -297,6 +297,7 @@
         vector_destroy(call_list);
     }
 cleanup:
+    routing_end();
     signal(SIGINT, callmgr_do_nothing);
     signal(SIGTERM, callmgr_do_nothing);
     close_inetsock(inet_sock, inetaddr);
