diff --exclude CVS --exclude .git -uNr u-boot-2009.11/cpu/lpc313x/u-boot.lds u-boot-2009.11.modified/cpu/lpc313x/u-boot.lds
--- u-boot-2009.11/cpu/lpc313x/u-boot.lds	2012-07-03 15:31:12.276631984 -0700
+++ u-boot-2009.11.modified/cpu/lpc313x/u-boot.lds	2012-07-03 15:29:43.278494405 -0700
@@ -61,6 +61,7 @@
 		drivers/spi/lpc313x_spi.o	(.text .rodata*)
 		drivers/mtd/spi/spi_flash.o	(.text .rodata*)
 		drivers/mtd/spi/atmel.o		(.text .rodata*)
+		drivers/mtd/spi/spansion.o	(.text .rodata*)
 #endif
 #ifdef CONFIG_CPU_USBDFU_BOOT
 		drivers/serial/usbtty.o       	(.text .rodata*)
diff --exclude CVS --exclude .git -uNr u-boot-2009.11/drivers/mtd/spi/spansion.c u-boot-2009.11.modified/drivers/mtd/spi/spansion.c
--- u-boot-2009.11/drivers/mtd/spi/spansion.c	2009-12-15 14:20:54.000000000 -0800
+++ u-boot-2009.11.modified/drivers/mtd/spi/spansion.c	2011-07-25 13:04:59.638841908 -0700
@@ -52,12 +52,14 @@
 #define SPSN_ID_S25FL128P	0x2018
 #define SPSN_EXT_ID_S25FL128P_256KB	0x0300
 #define SPSN_EXT_ID_S25FL128P_64KB	0x0301
+#define SPSN_EXT_ID_S25FL032P		0x4d00
 
 #define SPANSION_SR_WIP		(1 << 0)	/* Write-in-Progress */
 
 struct spansion_spi_flash_params {
 	u16 idcode1;
 	u16 idcode2;
+	u16 idmask2;
 	u16 page_size;
 	u16 pages_per_sector;
 	u16 nr_sectors;
@@ -79,6 +81,7 @@
 	{
 		.idcode1 = SPSN_ID_S25FL008A,
 		.idcode2 = 0,
+		.idmask2 = 0xffff,
 		.page_size = 256,
 		.pages_per_sector = 256,
 		.nr_sectors = 16,
@@ -87,6 +90,7 @@
 	{
 		.idcode1 = SPSN_ID_S25FL016A,
 		.idcode2 = 0,
+		.idmask2 = 0xffff,
 		.page_size = 256,
 		.pages_per_sector = 256,
 		.nr_sectors = 32,
@@ -95,6 +99,7 @@
 	{
 		.idcode1 = SPSN_ID_S25FL032A,
 		.idcode2 = 0,
+		.idmask2 = 0xffff,
 		.page_size = 256,
 		.pages_per_sector = 256,
 		.nr_sectors = 64,
@@ -103,6 +108,7 @@
 	{
 		.idcode1 = SPSN_ID_S25FL064A,
 		.idcode2 = 0,
+		.idmask2 = 0xffff,
 		.page_size = 256,
 		.pages_per_sector = 256,
 		.nr_sectors = 128,
@@ -111,6 +117,7 @@
 	{
 		.idcode1 = SPSN_ID_S25FL128P,
 		.idcode2 = SPSN_EXT_ID_S25FL128P_64KB,
+		.idmask2 = 0xffff,
 		.page_size = 256,
 		.pages_per_sector = 256,
 		.nr_sectors = 256,
@@ -119,11 +126,21 @@
 	{
 		.idcode1 = SPSN_ID_S25FL128P,
 		.idcode2 = SPSN_EXT_ID_S25FL128P_256KB,
+		.idmask2 = 0xffff,
 		.page_size = 256,
 		.pages_per_sector = 1024,
 		.nr_sectors = 64,
 		.name = "S25FL128P_256K",
 	},
+	{
+		.idcode1 = SPSN_ID_S25FL032A,
+		.idcode2 = SPSN_EXT_ID_S25FL032P,
+		.idmask2 = 0xff00,
+		.page_size = 256,
+		.pages_per_sector = 256,
+		.nr_sectors = 64,
+		.name = "S25FL032P",
+	},
 };
 
 static int spansion_wait_ready(struct spi_flash *flash, unsigned long timeout)
@@ -317,7 +334,7 @@
 	for (i = 0; i < ARRAY_SIZE(spansion_spi_flash_table); i++) {
 		params = &spansion_spi_flash_table[i];
 		if (params->idcode1 == jedec) {
-			if (params->idcode2 == ext_jedec)
+			if (params->idcode2 == (ext_jedec & params->idmask2))
 				break;
 		}
 	}
diff --exclude CVS --exclude .git -uNr u-boot-2009.11/include/configs/ea31xx.h u-boot-2009.11.modified/include/configs/ea31xx.h
--- u-boot-2009.11/include/configs/ea31xx.h	2011-07-25 21:44:26.946501979 -0700
+++ u-boot-2009.11.modified/include/configs/ea31xx.h	2011-07-25 18:15:54.209347529 -0700
@@ -202,6 +210,7 @@
 #define CONFIG_LPC313X_SPI		1
 #define CONFIG_SPI_FLASH		1
 #define CONFIG_SPI_FLASH_ATMEL		1
+#define CONFIG_SPI_FLASH_SPANSION       1
 #define CONFIG_CMD_SF			1
 #define CONFIG_ENV_IS_IN_SPI_FLASH	1
 
