diff --exclude CVS --exclude .git -uNr u-boot-2009.11/Makefile u-boot-2009.11.modified/Makefile
--- u-boot-2009.11/Makefile	2017-09-26 14:48:51.642618989 -0700
+++ u-boot-2009.11.modified/Makefile	2017-09-26 14:39:00.313734306 -0700
@@ -24,7 +24,7 @@
 VERSION = 2009
 PATCHLEVEL = 11
 SUBLEVEL =
-EXTRAVERSION = -mbari8
+EXTRAVERSION = -mbariA
 ifneq "$(SUBLEVEL)" ""
 U_BOOT_VERSION = $(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)
 else
diff --exclude CVS --exclude .git -uNr u-boot-2009.11/include/configs/ea31xx.h u-boot-2009.11.modified/include/configs/ea31xx.h
--- u-boot-2009.11/include/configs/ea31xx.h	2017-09-26 14:48:51.642618989 -0700
+++ u-boot-2009.11.modified/include/configs/ea31xx.h	2017-09-26 14:38:04.267159240 -0700
@@ -101,8 +101,8 @@
 #define CONFIG_NETMASK			255.255.255.0
 #endif
 
-/* Linux boot using network */
-#define CONFIG_BOOTCOMMAND		"run flashBoot"
+/* Linux boot using sdCard by default */
+#define CONFIG_BOOTCOMMAND		"run sdBoot"
 
 /*
  * Serial Driver Console
