diff -ru wpa_supplicant-2.5/src/utils/includes.h wpa_supplicant-2.5-builds/src/utils/includes.h
--- wpa_supplicant-2.5/src/utils/includes.h	2015-09-27 12:02:05.000000000 -0700
+++ wpa_supplicant-2.5-builds/src/utils/includes.h	2016-04-22 17:14:40.732269414 -0700
@@ -27,6 +27,7 @@
 #include <errno.h>
 #endif /* _WIN32_WCE */
 #include <ctype.h>
+#define isblank(c) ((c)<=' ')  //don't know why this is missing
 
 #ifndef _MSC_VER
 #include <unistd.h>
Only in wpa_supplicant-2.5-builds/wpa_supplicant: .config
